### ISSUE TITLE
The establishment of iptables rules lags behind Pod traffic occurrence

### ISSUE THREAD

**Issue by jimmy-zh** [cloudnativelabs/kube-router#873]
**What happened:**

In some cases, NPC's establishment of iptables rules lags behind Pod traffic occurrence, which results in network isolation failing to take effect. Maybe this is a knotty problem in consideration of K8s event synchronization mechanism.

**How to reproduce it:**
1. Deploy the following `Network Policy` in the K8s cluster:
```
apiVersion: extensions/v1beta1
kind: NetworkPolicy
metadata:
  name: test-np
  namespace: diagnosis
spec:
  egress:
  - to:
    - ipBlock:
        cidr: 14.215.0.0/16
  podSelector: {}
  policyTypes:
  - Egress
```
2. Deploy the following `Deployment` in the K8s cluster:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: diagnosis
  namespace: diagnosis
spec:
  replicas: 2
  selector:
    matchLabels:
      app: diagnosis
  template:
    metadata:
      labels:
        app: diagnosis
    spec:
      nodeName: 10.0.1.30
      containers:
      - name: diagnosis
        command: ["/bin/sh"]
        args: ["-c","ping 220.181.38.150"]
        image: jimmyzh/diagnosis:v1
```
3. Observe the logs of certain Pods:
```
kubectl logs diagnosis-66f959fd76-2wpln -n diagnosis | head

PING 220.181.38.150 (220.181.38.150) 56(84) bytes of data.
64 bytes from 220.181.38.150: icmp_seq=1 ttl=53 time=3.66 ms
64 bytes from 220.181.38.150: icmp_seq=2 ttl=53 time=3.57 ms
64 bytes from 220.181.38.150: icmp_seq=3 ttl=53 time=3.58 ms
```
4. Observe the iptables rules:
```
iptables-save -t filter

-A KUBE-POD-FW-3F2W5JGDV6M5LYIN -m comment --comment "rule for stateful firewall for pod" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A KUBE-POD-FW-3F2W5JGDV6M5LYIN -m comment --comment "run through nw policy test-np" -j KUBE-NWPLCY-ONWOLJXYLLQ2RYI2
-A KUBE-POD-FW-3F2W5JGDV6M5LYIN -m comment --comment "default rule to REJECT traffic destined for POD name:diagnosis-66f959fd76-2wpln namespace: diagnosis" -j REJECT --reject-with icmp-port-unreachable
```
5. Drop the stateful firewall rule manually:
```
iptables -t filter -D KUBE-POD-FW-3F2W5JGDV6M5LYIN -m comment --comment "rule for stateful firewall for pod" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
```
6. Observe the logs of the Pod above:
```
kubectl logs diagnosis-66f959fd76-2wpln -n diagnosis | tail

64 bytes from 220.181.38.150: icmp_seq=19347 ttl=53 time=3.59 ms
64 bytes from 220.181.38.150: icmp_seq=19348 ttl=53 time=3.58 ms
From 172.22.0.65 icmp_seq=19349 Destination Port Unreachable
From 172.22.0.65 icmp_seq=19350 Destination Port Unreachable
From 172.22.0.65 icmp_seq=19351 Destination Port Unreachable
```
The result indicates that the Pod traffic occurs before the establishment of iptables rules, and the related stateful firewall rule aggravates the ineffectiveness of isolation.

7. Add a 10 second sleep before the ping action in the `Deployment` above:
```
      containers:
      - name: diagnosis
        command: ["/bin/sh"]
        args: ["-c","sleep 10;ping 220.181.38.150"]
        image: jimmyzh/diagnosis:v1
```
8. Observe the logs of the new Pod:
```
kubectl logs diagnosis-796c7d859f-cpskr -n diagnosis | head

PING 220.181.38.150 (220.181.38.150) 56(84) bytes of data.
From 172.22.0.65 icmp_seq=1 Destination Port Unreachable
From 172.22.0.65 icmp_seq=2 Destination Port Unreachable
From 172.22.0.65 icmp_seq=3 Destination Port Unreachable
```
The result turns out to be normal after a sleep action.

**Comment by aauren** [comment]
@jimmy-zh I've suspected something similarly before. Just out of curiosity, would you mind adding a 60 second sleep before your ping to see what the results are?

If you're correct, I would expect the pings to fail from the beginning, as the iptables rules would have come into place before the first ping and restricted the traffic.

**Comment by jimmy-zh** [comment]
> @jimmy-zh I've suspected something similarly before. Just out of curiosity, would you mind adding a 60 second sleep before your ping to see what the results are?

@aauren No, certainly not. Actually, I've done similar tests before. I added two steps to the issue description to show the new result.

**Comment by jimmy-zh** [comment]
AFAIK most of Network Policy Controllers are implemented in a similar way:
1. Watch Network Policy related events, such as Pod Adding.
2. Synchronize iptables rules to achieve traffic filtering.

This mechanism does not seem to guarantee that iptables rules are established before the Pod runs. 

@murali-reddy @aauren Please tell me what you think about this problem. Thanks!

**Comment by murali-reddy** [comment]
If **--run-firewall=true** then we can change the default behaviour during the pod start to till network policies are evaluated for the pod is to drop traffic for both ingress and egress. perhaps this is right thing to do from security standpoint. Right now kube-router for a pod allows both ingress and egress traffic till network policies are evaluated for the pod which will decide to allow/drop both ingress and egress traffics.

But we run into the converse [problem](https://github.com/weaveworks/weave/issues/3464#issuecomment-443730137) where traffic from the pod is dropped which will be eventually permitted by network policies. However this is less of problem for TCP traffic  due to retry.

AFAIK there is no correct way to achive both the cases, as setting up the pod involes various asynchronous actions.

[pod “ready++”](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md) can be of some help in addressing some cases but can not be a general solution. Let me investigate bit further if there is better solution.

**Comment by aauren** [comment]
I'm going to close this issue for now.

There is still an open conversation around whether or not kube-router could / should drop all pod traffic until it is able to evaluate all of the possible network policies that might apply to the pod.

Right now kube-router still allows traffic to be open until the network policy is applied. Since we then allow established / related flows there is still the possibilities that some flows that might be otherwise disallowed might be allowed going forward if a traffic flow is established fast enough.

At some point in the future we may choose to change this default or to allow the user to configure this policy. However, for the time being, we are in a much better place with the advent of kube-router 1.2.X. Where once it used to take 1 - 5 minutes to setup network policy even in small - medium sized clusters, we are now below 5 seconds even for large clusters. And we are continuing to improve that.

Given that < 5 seconds is also about how long it takes to spin up a pod and for processes to start, I think that we are well within range of this no longer being a problem.

If someone runs into something down the road, feel free to ping and we can re-open this issue.

**Comment by morremeyer** [comment]
I am running into this issue from the opposite side: I run a k3s cluster that by default has a NetworkPolicy that only allow Egress traffic to coredns so that DNS resolution works.

For some tasks, I run CronJobs (e.g. database backups). The startup of those pods (from pod creation to the container running) oftentimes takes less than a second.

This is not enough time for kube-router to apply the iptables rules derived from the NetworkPolices that allow those (Cron)Job pods access to the database. Therefore, they fail to reach their targets and go into an Error state.

If policy application were faster or there was some possiblity to delay pod startup until the policies are applied, this case could/would be solved.

**Comment by aauren** [comment]
Unfortunately, Kubernetes doesn't really give us hooks into delaying pod start based upon CNI readiness. While I feel like we have gotten kube-router to implement policy pretty quickly, because of the nature of iptables, we are unlikely to be able to apply policy in less than a second in a consistent way on anything but the smallest of nodes.

I would recommend that you add a step that attempts to resolve DNS before starting your application and have that pause the launch of your database backup until DNS is resolve-able. If you wanted to separate out this functionality, you could also implement this functionality as a initContainer.

**Comment by morremeyer** [comment]
I'm thinking about implementing it as an initContainer, but it would make all of those jobs more complex. 

Anyway, the situation here is clear, thanks for the feedback.

**Comment by disconn3ct** [comment]
> If someone runs into something down the road, feel free to ping and we can re-open this issue.

I'm running a k3s cluster on RPI and getting a reproducible delay of at least 45s before the rules are applied. Does that count as `something`?

I'm happy to gather any info that can help. I really need networkpolicies to work reliably and I don't want to go back to Calico. (I saw above that it isn't a trivial fix, but was the default-deny flag ever exposed? I am fine with blocking connections until the correct rules are applied.)

**Comment by aauren** [comment]
@disconn3ct 45 seconds seems pretty large, I definitely wouldn't expect that. Can you post the version of kube-router that you are running?

Additionally, it would help if you could enable verbose logging with `-v 2` and give us the logs for an entire full policy sync.

The relevant logs will be bookend-ed with something like the following:
```
Starting sync of iptables with version: <version>
...
sync iptables took <time>
```

**Comment by disconn3ct** [comment]
I'm using the version embedded with k3s v1.24.9+k3s1
`Starting the netpol controller version v1.5.2-0.20221026101626-e01045262706, built on 2022-12-21T00:28:07Z, go1.18.9`

I can't find any way to pass arguments to it through k3s. Do you know of a way? 

Unfortunately I missed part of your earlier message though - so long as the established/related problem isn't solved, kube-router can't meet my needs. Security via race conditions make me very nervous. Sorry to resurface an old bug for nothing.

**Comment by aauren** [comment]
Unfortunately, I don't know how to pass arguments through k3s to kube-router, you'd have to follow up with upstream k3s on how to do that.

I agree with you about security via race conditions being un-ideal. I think that the comment here still applies: https://github.com/cloudnativelabs/kube-router/issues/873#issuecomment-898507438 . If you're willing, please open a new issue describing your use-case for added security, reference this issue, and ask for a new feature to allow optionally changing the default posture of kube-router from ACCEPT to DENY until network policy is applied.

I don't think that this would be a big change, and it is something that we could likely put in for our upcoming 2.0.0 release of kube-router.

**Comment by Perdjesk** [comment]
Similar as reported in https://github.com/k3s-io/k3s/issues/13432 k3s repository, with reproducible steps that demonstrate that applied network policy filtering is delayed at pod startup.

Eventual connectivity can be tolerated, but eventual isolation/protection/filtering shouldn't (in the case of network policy already in-place and newly created pods).

The current implementation of kube-router for NetworkPolicies is not conformant to the docs. 

> Once the NetworkPolicy is handled by a network plugin,
>
> 1. All newly created pods affected by a given NetworkPolicy will be isolated before they are started.
>   Implementations of NetworkPolicy must ensure that filtering is effective throughout
>   the Pod lifecycle, even from the very first instant that any container in that Pod is started.
>   Because they are applied at Pod level, NetworkPolicies apply equally to init containers,
>   sidecar containers, and regular containers.

https://github.com/kubernetes/website/blob/096fbdd861b94d70748246ddfc323a44d4f37f77/content/en/docs/concepts/services-networking/network-policies.md?plain=1#L374-L380

**Comment by aauren** [comment]
Hey @Perdjesk, first off commenting on an issue that is about 3 years old is not really a great way to get support around here.

I can understand that you are frustrated, however, this is nothing that kube-router has control over without extending beyond the realm of what we are able to support. The flow for how containers start is as follows:

1. kube-scheduler chooses a node to schedule a pod on 
2. kubelet on the node sees the scheduled pod for its node
3. kubelet sends the containers in the pod to the container runtime via CRI
4. the CRI calls the CNI hooks and the pod is assigned an IP and the sandbox is created
5. kubelet starts the container and lets the kube-apiserver know the pod IP address
6. kube-router sees the pod and its IP address and begins executing hooks one of those hooks is applying networking policy

As you can see, and as @murali-reddy points out above, there is no hook that allows kube-router to tell Kubernetes to block until network policies can apply. Additionally, as network policies require layer 3 networking, there is a chicken and egg condition here as well. kube-router needs the pod IP so that it can apply network policy, and the policy needs to be applied before the pod IP is given in order to ensure that no traffic can be sent.

Some more complex projects have ways to work around this. Both Cilium and Calico for instance own the CNI layer and integrate in complex ways between the CNI and the dataplane of the network provider. kube-router however, relies on the generic upstream CNI IPAM plugins: https://github.com/containernetworking/plugins/tree/main/plugins/main

It is beyond the scope of the kube-router project to integrate at this layer and that is unlikely to change with the current state of the project. I would recommend that you either work around this limitation, via something like an init container or a sleep or some such. Or I would suggest that you use a different network provider that can provide what you're looking for.

**Comment by Perdjesk** [comment]
> Hey [@Perdjesk](https://github.com/Perdjesk), first off commenting on an issue that is about 3 years old is not really a great way to get support around here.

Sorry if that bothered you. A general accepted rule is as well not to open a new issue for something already reported and still relevant.

> I can understand that you are frustrated,

Unnecessary emotional projection.

> this is nothing that kube-router has control over without extending beyond the realm of what we are able to support. 

In a previous answer you were suggesting that the default policy could be changed. If the default was changed to a deny policy until pod's network setup and netpol application, it would allow eventual connectivity and guarantee initial isolation. The current default of open connectivity avoids startup errors for most users, but the initial enforcement gap is easy to overlook as it rarely causes visible failures.

The current default prioritizes connectivity over isolation, trading network policy guarantees for fewer startup failures.
Network policies principal purpose are to filter traffic, silently allowing all traffic at pod startup defeats this purpose. 

This limitation should be made more visible to users in documentation.
The default could be deny rather than allow, as originally considered, to honor the purpose of network policies.

Note:
[This previous comment](https://github.com/cloudnativelabs/kube-router/issues/873#issuecomment-1051268779 ), which is a startup failure, is the inverse of the original reported behavior, which is surprising as the connectivity should be open at pod's startup and not causes errors. It might be another issue (e.g the pod network setup at pod's creation) and not related to kube-router network policies.

**Comment by aauren** [comment]
After talking about it internally with other maintainers, we decided that there is a path forward for optionally allowing people to opt-into a deny-by-default posture without breaking other workloads on the node and without diving into a bespoke CNI implementation that would not be supportable by the project.

It has trade-offs, but these are trade-offs that cluster administrators can opt into via a new CLI parameter. More details on the implementation can be found on the associated PR #2079
